compiler/src/yul/abi/utils.rs

```diff
@@ -0,0 +1,73 @@
+use fe_semantics::namespace::types::{
+    AbiArraySize,
+    AbiDecodeLocation,
+    AbiEncoding,
+    AbiType,
+    AbiUintSize,
+};
+use yultsur::*;
+
+pub fn head_offsets<T: AbiEncoding>(types: &[T]) -> (Vec<usize>, usize) {
+    let mut offsets = vec![];
+    let mut curr_offset = 0;
+
+    for typ in types {
+        offsets.push(curr_offset);
+
+        curr_offset += match typ.abi_type() {
+            AbiType::Array {
+                size: AbiArraySize::Dynamic { .. },
+                ..
+            } => 32,
+            AbiType::Array {
+                size: AbiArraySize::Static { size },
+                inner,
+            } => match *inner {
+                AbiType::Array { .. } => unimplemented!(),
+                AbiType::Uint {
+                    size: AbiUintSize { padded_size, .. },
+                } => ceil_32(padded_size * size),
+            },
+            AbiType::Uint {
+                size: AbiUintSize { padded_size, .. },
+            } => padded_size,
+        };
+    }
+
+    (offsets, curr_offset)
+}
+
+/// Generates an ABI encoding function name for a given set of types.
+pub fn encode_name<T: AbiEncoding>(types: &[T]) -> yul::Identifier {
+    let mut full_name = "abi_encode".to_string();
+
+    for typ in types {
+        full_name.push('_');
+        full_name.push_str(&typ.abi_safe_name());
+    }
+
+    identifier! { (full_name) }
+}
+
+/// Generates an ABI decoding function name for a given type.
+pub fn decode_name<T: AbiEncoding>(typ: &T, loc: AbiDecodeLocation) -> yul::Identifier {
+    let mut full_name = "abi_decode".to_string();
+    let loc = match loc {
+        AbiDecodeLocation::Memory => "mem",
+        AbiDecodeLocation::Calldata => "calldata",
+    };
+    full_name.push('_');
+    full_name.push_str(loc);
+    full_name.push('_');
+    full_name.push_str(&typ.abi_safe_name());
+
+    identifier! { (full_name) }
+}
+
+pub fn ceil_32(n: usize) -> usize {
```

Comment:
use the simpler 1-line expression thats used in the runtime version of this function